include/ctrace/ctrace.h

```diff
@@ -0,0 +1,283 @@
+#  else
+#   define CTRACE_C11_FEATURES OFF
+#  endif
+# else
+#  define CTRACE_C11_FEATURES OFF
+# endif
+#else
+# define CTRACE_C11_FEATURES OFF
+#endif
+
+#if defined(__cplusplus)
+# define CTRACE_NORETURN
+#elif CTRACE_C11_FEATURES == ON
+# include <stdnoreturn.h>
+# define CTRACE_NORETURN noreturn
+#else
+# define CTRACE_NORETURN
+#endif
+
+#if defined(_WIN32) || defined(__CYGWIN__)
+# define CTRACE_API C_LINKAGE __declspec(dllexport)
+#else
+# define CTRACE_API C_LINKAGE
+#endif
+
+#define ON 1
+#define OFF 0
+
+#if !defined(CTRACE_EXCEPTIONS)
+# define CTRACE_EXCEPTIONS OFF
+#endif
+
+#if !defined(CTRACE_FREESTANDING)
+# if defined(__STDC_HOSTED__)
+#  if __STDC_HOSTED__ == 1
+#   define CTRACE_FREESTANDING OFF
+#  else
+#   define CTRACE_FREESTANDING ON
+#  endif
+# else
+#  define CTRACE_FREESTANDING OFF
+# endif
+#endif
+
+#define CTRACE_BITCOUNT(x) (sizeof(x) * CHAR_BIT)
+#define EFLAG(n) 1 << n
+
+#define ESC     "\033["
+#define RESET   ESC "0m"
+#define RED     ESC "31m"
+#define GREEN   ESC "32m"
+#define YELLOW  ESC "33m"
+#define BLUE    ESC "34m"
+#define MAGENTA ESC "35m"
+#define CYAN    ESC "36m"
+
+BEGIN_CTRACE
+    typedef enum { cfalse, ctrue } ctrace_boolean;
+    typedef const char *ctrace_format_t;
+
+    // general::string:
+    typedef struct {
+        const void *data;
+        unsigned type;
+    } ctrace_string;
+
+    struct ctrace_buffered_string;
+    struct ctrace_exception_ctx;
+
+    /* Signifies `ctrace_free_string` should not be called.
+     */
+    typedef ctrace_string ctrace_managed_string;
+
+    /* Signifies a buffered string is required for effects.
+     */
+    typedef ctrace_string ctrace_mut_string;
+
+    CTRACE_API void ctrace_clear_string(ctrace_mut_string string);
+    CTRACE_API void ctrace_write_string(ctrace_mut_string to, ctrace_string from);
+    CTRACE_API void ctrace_write_move_string(ctrace_mut_string to, ctrace_string from);
+    CTRACE_API void ctrace_format_string(ctrace_mut_string to, ctrace_format_t, ...);
+
+    /* Allocates a buffer of len + 1.
+     */
+    CTRACE_API ctrace_mut_string ctrace_make_buffered_string(unsigned len);
+    CTRACE_API const char *ctrace_get_cstring(ctrace_string string);
+    CTRACE_API const char *ctrace_get_string_type(ctrace_string string);
+    CTRACE_API const char *ctrace_get_underlying_string_type(ctrace_string string);
+    CTRACE_API unsigned ctrace_get_string_size(ctrace_string string);
+    CTRACE_API void ctrace_free_string(ctrace_string string);
+    CTRACE_API int ctrace_fputs(ctrace_string string, FILE *f);
+    CTRACE_API int ctrace_puts(ctrace_string string);
+
+    // api:
+    typedef void *ctrace_frame_types;
+
+    typedef struct {
+        ctrace_frame_types data;
+        struct {
+            unsigned type : CTRACE_BITCOUNT(unsigned) - 1;
+            ctrace_boolean resolved : 1;
+        };
+    } ctrace_wrapped_trace,
+        ctrace_raw_trace,
+        ctrace_object_trace,
+        ctrace_trace;
```

Comment:
It'd be more type safe to have different types for each type of trace, and more consistent with the C++ api. It would be nice to have `ctrace_trace trace; puts(trace.frames[0].filename);` style access rather than have these as opaque types.